libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/RouteComparator.java

```diff
@@ -0,0 +1,73 @@
+package com.mapbox.services.android.navigation.ui.v5;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import com.mapbox.api.directions.v5.models.DirectionsResponse;
+import com.mapbox.api.directions.v5.models.DirectionsRoute;
+import com.mapbox.api.directions.v5.models.RouteLeg;
+
+import java.util.List;
+
+class RouteComparator {
+
+  private static final int FIRST_ROUTE = 0;
+  private static final int ONE_ROUTE = 1;
+  private final NavigationViewRouter navigationViewRouter;
+
+  RouteComparator(NavigationViewRouter navigationViewRouter) {
+    this.navigationViewRouter = navigationViewRouter;
+  }
+
+  void compare(@NonNull DirectionsResponse response, @Nullable DirectionsRoute chosenRoute) {
+    if (isValidRoute(response)) {
+      List<DirectionsRoute> routes = response.routes();
+      DirectionsRoute bestRoute = routes.get(FIRST_ROUTE);
+      if (isNavigationRunning(chosenRoute)) {
+        bestRoute = obtainMostSimilarRoute(routes, bestRoute, chosenRoute);
+      }
+      navigationViewRouter.updateCurrentRoute(bestRoute);
+    }
+  }
+
+  private DirectionsRoute obtainMostSimilarRoute(List<DirectionsRoute> routes, DirectionsRoute currentBestRoute,
+                                                 DirectionsRoute chosenRoute) {
+    DirectionsRoute mostSimilarRoute = currentBestRoute;
+    if (routes.size() > ONE_ROUTE) {
+      mostSimilarRoute = findMostSimilarRoute(chosenRoute, routes);
+    }
+    return mostSimilarRoute;
+  }
+
+  private DirectionsRoute findMostSimilarRoute(DirectionsRoute chosenRoute, List<DirectionsRoute> routes) {
+    int routeIndex = 0;
+    String chosenRouteLegDescription = obtainRouteLegDescriptionFrom(chosenRoute);
+    int minSimilarity = Integer.MAX_VALUE;
+    for (int index = 0; index < routes.size(); index++) {
```

Comment:
You could make this a for each (`for(DirectionsRoute route: routes)`) and instead of `routeIndex` keep track of the route